mirgecom/profiling.py

```diff
@@ -0,0 +1,202 @@
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+from meshmode.array_context import PyOpenCLArrayContext
+import pyopencl as cl
+from pytools.py_codegen import PythonFunctionGenerator
+import loopy as lp
+import numpy as np
+from dataclasses import dataclass
+import pytools
+
+__doc__ = """
+.. autoclass:: PyOpenCLProfilingArrayContext
+"""
+
+
+@dataclass
+class ProfileResult:
+    """Class to hold the results of a single kernel execution."""
+    time: int
+    flops: int
+    bytes_accessed: int
+    footprint_bytes: int
+
+
+@dataclass
+class ProfileEvent:
+    """Class to hold a profile event that has not been seen by the profiler yet."""
+    cl_event: cl._cl.Event
+    program: lp.kernel.LoopKernel
+    kwargs: dict
+
+
+class PyOpenCLProfilingArrayContext(PyOpenCLArrayContext):
+    """An array context that profiles kernel executions.
+
+    .. automethod:: table_profiling_data
+    .. automethod:: call_loopy
+
+    Inherits from :class:`meshmode.array_context.PyOpenCLArrayContext`."""
+
+    def __init__(self, queue, allocator=None) -> None:
+        super().__init__(queue, allocator)
+
+        if not queue.properties & cl.command_queue_properties.PROFILING_ENABLE:
+            raise RuntimeError("Profiling was not enabled in the command queue. "
+                 "Please create the queue with "
+                 "cl.command_queue_properties.PROFILING_ENABLE.")
+
+        self.profile_events = []
+        self.profile_results = {}
+        self.kernel_stats = {}
+
+    def _finish_profile_events(self) -> None:
+        # First, wait for event completion
+        if self.profile_events:
+            cl.wait_for_events([pevt.cl_event for pevt in self.profile_events])
+
+        for t in self.profile_events:
+            program = t.program
+            res = self._get_kernel_stats(program, t.kwargs)
+            res.time = t.cl_event.profile.end - t.cl_event.profile.start
+
+            self.profile_results.setdefault(program, []).append(res)
+
+        self.profile_events = []
+
+    def table_profiling_data(self) -> pytools.Table:
```

Comment:
```suggestion
    def tabulate_profiling_data(self) -> pytools.Table:
```
https://www.merriam-webster.com/dictionary/table (verb)

Reply:
Done.